hw/ip/boot_rom/boot_rom.S

```diff
@@ -2,47 +2,156 @@
+
+_copy_from_flash:
+       lui    a1, SPI_HOST_START_ADDRESS_20bit
+       // Enable spi
+       lui    a0, 0x80000
+       lw     a4, SPI_HOST_CONTROL_REG_OFFSET(a1)
+       or     a4, a4, a0
+       sw     a4, SPI_HOST_CONTROL_REG_OFFSET(a1)
+       // Create config_opt: 0x0FFF0001
+       lui    a4, 0xfff0
+       addi   a4, a4, 1
+       sw     a4, SPI_HOST_CONFIGOPTS_0_REG_OFFSET(a1)
+       // Set spi csid
+       li     a0, 0
+       sw     a0, SPI_HOST_CSID_REG_OFFSET(a1)
+       // Set spi receiver watermark to 8
+       lw     a4, SPI_HOST_CONTROL_REG_OFFSET(a1)
+       andi   a4, a4, -256 # ~SPI_HOST_CONTROL_RX_WATERMARK_MASK
+       ori    a4, a4, 8
+       sw     a4, SPI_HOST_CONTROL_REG_OFFSET(a1)
+
+       // Power up flash (0xab flash command)
+       li     a4, 0xab
+       sw     a4, SPI_HOST_DATA_REG_OFFSET(a1)
+       // Power up command: 0x00002003 (transmit 4 words)
+       lui    a4, 0x2
+       addi   a4, a4, 3
+       sw     a4, SPI_HOST_COMMAND_REG_OFFSET(a1)
+
+_wait_spi_ready_cmd_pwr:
+       // Wait spi is ready for next command
+       lw     a4, SPI_HOST_STATUS_REG_OFFSET(a1)
+       bgez   a4, _wait_spi_ready_cmd_pwr
+
+       // Fill TX FIFO with TX data (read command 0x03 + 3B address 0x000)
+       li     a4, 0x03
+       sw     a4, SPI_HOST_DATA_REG_OFFSET(a1)
+
+_wait_spi_ready_tx_init:
+       nop    # otherwise ready bit check is too fast
+       lw     a4, SPI_HOST_STATUS_REG_OFFSET(a1)
+       bgez   a4, _wait_spi_ready_tx_init
+       // Read command: 0x00002203
+       lui    a4, 0x2
+       addi   a4, a4, 515 # spi cmd: txonly + stdspeed + csaat + 4B
+       sw     a4, SPI_HOST_COMMAND_REG_OFFSET(a1)
+
+_wait_spi_ready_read_prog:
+       nop    # otherwise ready bit check is too fast
+       lw     a5, SPI_HOST_STATUS_REG_OFFSET(a1)
+       bgez   a5, _wait_spi_ready_read_prog
+       li     a2, 0 # src ptr (flash)
+       li     a3, 1024 # 1KiB copy size
+       li     s1, 0 # dst ptr (ram)
+
+       // For loop until the 1KiB copy from flash to ram is done
+       // 256-bytes copies are done
+       li     s6, 256
+       lui    s0, 0x1
+       addi   s5, s0, 767 # spi cmd: rxonly + stdspeed + csaat + 255 bytes
+
+_32B_chunk_loop:
+       blt    s6, a3, _read_32B_chunk
+       // End the transaction if last 256 bytes
+       addi   s5, s0, 255 # spi cmd: rxonly + stdspeed + 255 bytes
+
+_read_32B_chunk:
+       sw     s5, SPI_HOST_COMMAND_REG_OFFSET(a1)
+
+_wait_spi_ready_read_32B_chunk:
+       nop    # otherwise ready bit check is too fast
+       lw     a5, SPI_HOST_STATUS_REG_OFFSET(a1)
+       bgez   a5, _wait_spi_ready_read_32B_chunk
+       addi   s7, s1, 256 # add 32*8 (256 bytes) to dst ptr
+
+_wait_spi_rxwm_8_words:
+       lw     a5, SPI_HOST_STATUS_REG_OFFSET(a1)
+       srli   a5, a5, 0x14
+       andi   a5, a5, 1 # SPI_HOST_STATUS_RXWM_BIT
+       beqz   a5, _wait_spi_rxwm_8_words
+       // Read 32B (8 words) from RX FIFO
+       lw     a7, SPI_HOST_DATA_REG_OFFSET(a1)
```

Comment:
although less efficient, a loop would decrease the ROM size, especially as these LW and SW are not compressed